### ISSUE TITLE
Cannot host CoreWcf in IIS under ASP.NET Core 3.1

### ISSUE THREAD

**Issue by allderidge** [CoreWCF/CoreWCF#313]
I cannot find a way of hosting a CoreWcf HTTP SOAP service in IIS or IISExpress on versions of ASP.NET Core later than 2.1. Please close the bug if support for IIS on later dot net versions is already planned, thanks.

## Repo Steps

In Visual Studio:
* Create a new 'ASP.NET Core Web Application'
* 'ASP.NET Core 3.1' 
* Select the 'Empty' option
* Add the CoreWcf Nuget packages
* Add services.AddServiceModelServices(); in ConfigureServices

## Error raised

The exception 'Application is running inside IIS process but is not configured to use IIS serve' is raised running the code:

```
System.InvalidOperationException
  HResult=0x80131509
  Message=Application is running inside IIS process but is not configured to use IIS server.
  Source=Microsoft.AspNetCore.Server.IIS
  StackTrace:
   at Microsoft.AspNetCore.Server.IIS.Core.IISServerSetupFilter.<>c__DisplayClass2_0.<Configure>b__0(IApplicationBuilder app)
   at Microsoft.AspNetCore.HostFilteringStartupFilter.<>c__DisplayClass0_0.<Configure>b__0(IApplicationBuilder app)
   at Microsoft.AspNetCore.Hosting.GenericWebHostService.<StartAsync>d__31.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Microsoft.Extensions.Hosting.Internal.Host.<StartAsync>d__9.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.<RunAsync>d__4.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.<RunAsync>d__4.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run(IHost host)
   at IisCoreWcfTest.Program.Main(String[] args) in C:\Users\jallderidge\source\repos\CoreWcf_IisHostBugReport\IisCoreWcfTest\Program.cs:line 11

  This exception was originally thrown at this call stack:
    Microsoft.AspNetCore.Server.IIS.Core.IISServerSetupFilter.Configure.AnonymousMethod__0(Microsoft.AspNetCore.Builder.IApplicationBuilder)

Microsoft.AspNetCore.HostFilteringStartupFilter.Configure.AnonymousMethod__0(Microsoft.AspNetCore.Builder.IApplicationBuilder)
    Microsoft.AspNetCore.Hosting.GenericWebHostService.StartAsync(System.Threading.CancellationToken)
    System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
    System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
    System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
    System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
    Microsoft.Extensions.Hosting.Internal.Host.StartAsync(System.Threading.CancellationToken)
    System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
    System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
```

## Example Code

```
    public class Program
    {
        public static void Main(string[] args) => CreateHostBuilder(args).Build().Run();

        public static IHostBuilder CreateHostBuilder(string[] args) =>
            Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(webBuilder =>
                {
                    webBuilder.UseStartup<Startup>();
                });
    }

    public class Startup
    {
        public void ConfigureServices(IServiceCollection services)
        {
            // adding this line will prevent startup in IIS express
            services.AddServiceModelServices();
        }

        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
        }
    }
```

An example solution is at [https://github.com/allderidge/CoreWcf_IisHostBugReport](https://github.com/allderidge/CoreWcf_IisHostBugReport)

**Comment by inforithmics** [comment]
I had the same problem changing the code with the code below and the exception doesn't occour anymore.

```
public class Program
    {
        public static void Main(string[] args) => CreateHostBuilder(args).Build().Run();

        public static IHostBuilder CreateHostBuilder(string[] args) =>
            Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(webBuilder =>
                {
                    webBuilder.UseStartup<Startup>();
                });
    }

    public class Startup
    {
        public void ConfigureServices(IServiceCollection services)
        {
            // adding this line will prevent startup in IIS express
            services.AddServiceModelServices();
            services.FixWcfCore();
        }

        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
        }
    }

    public static class CoreWcfHelper
    {
        public static void FixWcfCore(this IServiceCollection services)
        {
            for (int i = 0; i < services.Count; i++)
            {
                if (services[i].ServiceType == typeof(IStartupFilter))
                {
                    var implType = services[i]?.ImplementationInstance?.GetType();
                    if (implType?.Name == "IISServerSetupFilter")
                    {
                        var field = implType.GetField(@"_virtualPath", System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.NonPublic);
                        var virtualPath = field?.GetValue(services[i].ImplementationInstance);

                        if (virtualPath is string virtualPathString)
                        {
                            var iisServerSetupFilter = new IISServerSetupFilter(virtualPathString);
                            services[i] = new ServiceDescriptor(services[i].ServiceType, iisServerSetupFilter);
                        }
                    }
                }
            }
        }
    }

    public sealed class IISServerSetupFilter : IStartupFilter
    {
        private readonly string virtualPath;

        public IISServerSetupFilter(string virtualPath) => this.virtualPath = virtualPath;

        public System.Action<IApplicationBuilder> Configure(System.Action<IApplicationBuilder> next)
        {
            return app =>
            {
                app.UsePathBase(this.virtualPath);
                next(app);
            };
        }
    }
```

The reason that an exception is thrown is because the original IISServerSetupFilter  has following code in it

```
if (app.ApplicationServices.GetService<IServer>()?.GetType() != typeof (IISHttpServer))
          throw new InvalidOperationException("Application is running inside IIS process but is not configured to use IIS server.");
```

And because CoreWcf replaces IServer implementation it isn't anymore IISHttpServer and therefore this filter throws an exception.

The solution is to replace the IISServerSetupFilter with the custom implementation that hasn't this check, and it works without problems.

**Comment by mconnew** [comment]
@inforithmics, thanks for working out the problem and providing a workaround. There's also a problem with CoreWCF presuming the base path is going to be / which breaks when using HttpSys. I suspect that will also cause problems for IIS hosting. I still need to work out how to fix that one but I'll ship a servicing release with that fix at the same time as this as.

**Comment by davidfowl** [comment]
I'll take a look as well

**Comment by mconnew** [comment]
I've been thinking about this and I think the solution is to not do the wrapping when IIS hosted. The purpose of the wrapping is because there's no other way to cause the host start to fail if there's some problem with how CoreWCF has been configured. This is because we have to wait until the HTTP host has started before we can finish initializing CoreWCF. But this doesn't make sense when IIS hosted as you don't have a Main entry point and a console to see the exception being thrown. So the benefit of doing this is lost when using IIS.

**Comment by mconnew** [comment]
Reopening because I've found some other issues.

**Comment by allderidge** [comment]
IIS and IISExpress work when running with the latest code base built locally both with the sample and an existing application server migrated over to net5.0 . The converted application server is stable and all the external API tests pass so CoreWcf looks to be completely functional thanks.

Note that on 0.1.0 and 0.2.0-alpha with the fix provided above hosting on IISExpress works but I cannot host on IIS due to the following exception being generated (seen when attaching a debugger):

'The message with To 'http://localhost/iiscorewcftest/basichttp' cannot be processed at the receiver, due to an AddressFilter mismatch at the EndpointDispatcher.  Check that the sender and receiver's EndpointAddresses agree.'

thrown from: CoreWCF.Primitives.dll!CoreWCF.Dispatcher.ChannelHandler.ReplyFailureAsync(CoreWCF.Channels.RequestContext request, CoreWCF.Dispatcher.ChannelHandler.RequestInfo requestInfo, CoreWCF.Channels.Message fault, string action, string reason, CoreWCF.FaultCode code)

The exception does not propagate outside IIS correctly; an Accepted:202 http response code is returned with an empty response message.

**Comment by inforithmics** [comment]
My workaround isn't needed anymore with the newest CoreWCf nuget packages.

**Comment by allderidge** [comment]
F.Y.I. after testing with IIS there is an measure you must take in order to get CoreWcf working with Asp.Net Controllers, this is to invoke UseRouting() before UseServiceModel() on the IApplicationBuilder. If you don't do this then all routed requests for Controllers result in a 404. Note that I've only tested this with asp.netcore 5.0 .

E,g,:
```
public void Configure(IApplicationBuilder app)
{
    app.UseRouting();
    app.UseServiceModel(builder =>
    {
        builder
            .AddService<EchoService>()
            .AddServiceEndpoint<EchoService, IEchoService>(new BasicHttpBinding(), "/basichttp");
    });
    app.UseEndpoints(endpoints =>
    {
        endpoints.MapControllers();
    });
}
```

In testing this isn't the case with Kestral or IISExpress but these typically have an empty root address which I'm guessing makes the difference. I.e. app.UseServiceModel somehow causes routing to fail for the url http://localhost/IisCoreWcfTest/api/Test but succeed for http://localhost:5000/api/Test.

I've also had to the following for later versions of asp.net to prevent the `'Synchronous operations are disallowed. Call ReadAsync or set AllowSynchronousIO to true'` exception being thrown:

```
services.Configure<IISServerOptions>(options => // and or the same call with KestrelServerOptions
{
    options.AllowSynchronousIO = true;
});
```

The exception is coming from HttpRequestContext.cs line 429, which has a the comment `'// TODO: Look into useing PipeReader with look-ahead'` above it indicating its a known issue.

**Comment by davidfowl** [comment]
Wcf should start using endpoint routing. It would solve the ordering problem nicely

**Comment by mconnew** [comment]
@davidfowl, if you support .NET Framework with asp.net core 5.0, then I would do that asap. Right now the focus is on getting some core features working on .NET Framework which means staying with asp.net core 2.1 for now. Once we have a few key features implemented, then we'll more to the latest version. The main one is WSDL support.

**Comment by davidfowl** [comment]
CoreWcf can cross compile and offer the new API when targeting .net core 3.1 and up

**Comment by mconnew** [comment]
ASP.NET Core 2.1 runs perfectly fine on .NET Core 3.1. I don't know of a way to to make those api's only available when using ASP.NET Core 3.1. I suspect it will require some changes internal to CoreWCF to switch to the endpoint way of doing things. I can't see a clean way to do this without a lot of hacks as I can't make the code reference 3.1 libraries even on .NET Core 3.1 because of ASP.NET Core 2.1.

**Comment by davidfowl** [comment]
You target both 2.1 and 3.1 the you ifdef. It's not too bad, lots of libraries do it

**Comment by g7ed6e** [comment]
Hi. First thanks for this great job it will definitely open legacy workload migrations on containers.
Just a question why not being only focused on the latest ASP.NET core LTS i.e 3.1 and drop ASP.NET core 2.1 support. The whole goal is to support WCF on latest .NET core i think and I don't see right now the production use case of running CoreWCF on top of .NET full fwk.

**Comment by Kralizek** [comment]
@g7ed6e i guess the idea is to enable step by step migrations.

Wcf on windows
CoreWcf on windows/netfx
CoreWcf on windows/netcore
CoreWcf on linux/netcore

**Comment by cszczotka** [comment]
Hi,
@inforithmics @allderidge  do you run CoreWCF on IIS ( inprocess hosting model )  ? Do you have any example how to set up ?  I have similar issue with version 0.1.0 like @allderidge and new issue with 0.2.0 ( see below exception stack ) for .net core 3.1.  

`_2021-08-04 12:12:30.0494|10|-|ERROR|Microsoft.AspNetCore.Server.IIS.Core.IISHttpContext+Log.ApplicationError|Connection ID "12105675799982544576", Request ID "80009a9a-0000-a800-b63f-84710c7967bb": An unhandled exception was thrown by the application.Exception class: UriFormatException
Message: Invalid URI: The hostname could not be parsed.
Source: System.Private.Uri
Method: Void CreateThis(System.String, Boolean, System.UriKind)
Stack Trace:    at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)
   at System.Uri..ctor(String uriString)
   at CoreWCF.ServiceHostObjectModel`1.FixUri(String address)
   at CoreWCF.ServiceHostObjectModel`1.ApplyConfiguration()
   at CoreWCF.ServiceHostBase.InitializeDescription(UriSchemeKeyedCollection baseAddresses)
   at CoreWCF.ServiceHostObjectModel`1..ctor(IServiceProvider serviceProvider, ServiceBuilder serviceBuilder, ILogger`1 logger)
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at CoreWCF.Description.DispatcherBuilder.BuildDispatcher[TService](ServiceConfiguration`1 serviceConfig, IServiceProvider services)
   at CoreWCF.Configuration.ServiceConfiguration`1.GetDispatchers()
   at CoreWCF.Configuration.DispatcherBuilderImpl.BuildDispatchers(Type serviceType)
   at CoreWCF.Channels.ServiceModelHttpMiddleware.BuildBranch()
   at CoreWCF.Channels.ServiceModelHttpMiddleware.EnsureBranchBuilt()
   at CoreWCF.Channels.ServiceModelHttpMiddleware.BuildBranchAndInvoke(HttpContext request)
   at CoreWCF.Channels.ServiceModelHttpMiddleware.InvokeAsync(HttpContext context)
   at Microsoft.AspNetCore.Builder.Extensions.UsePathBaseMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Builder.Extensions.UsePathBaseMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContextOfT`1.ProcessRequestAsync()  (Depth: 0)
_`

**Comment by allderidge** [comment]
@cszczotka I have a example of an echo service hosted in IIS using CoreWCF 0.2.0 and AspNetCore 5.0 which you can get via https://github.com/allderidge/CoreWcf_IisHostBugReport/tree/Test_0.2.0_With_Service .

The example has a service project 'IisCoreWcfTest' and a client project 'Client'. Debug the service project and then debug the client project to invoke the WCF service.

There are three launch settings for IIS, IISExpress and Console which I have all had working. You need to switch the URL in the client Program.cs to target the chosen launch setting.

**Comment by mconnew** [comment]
@cszczotka, are you using + or * in your url?

**Comment by mconnew** [comment]
I believe this scenario is working now. If that's not the case, please reopen.
